aten/src/ATen/core/IList_inl.h

```diff
@@ -0,0 +1,215 @@
+) {
+  return it.payload_.unboxed_iterator;
+}
+
+inline
+c10::detail::IListConstRef<at::Tensor>
+TORCH_ILIST_IMPL(at::Tensor, Unboxed)::iterator_get(
+    const list_type::const_iterator& it
+) {
+  return *it;
+}
+
+inline
+const TORCH_ILIST_IMPL(at::Tensor, Boxed)::list_type&
+TORCH_ILIST_IMPL(at::Tensor, Boxed)::unwrap(
+    const c10::IList<at::Tensor>& ilist
+) {
+  return *ilist.payload_.boxed;
+}
+
+inline
+TORCH_ILIST_IMPL(at::Tensor, Boxed)::list_type::const_iterator&
+TORCH_ILIST_IMPL(at::Tensor, Boxed)::unwrap(
+    c10::IListIterator<at::Tensor>& it
+) {
+  return it.payload_.boxed_iterator;
+}
+
+inline
+const TORCH_ILIST_IMPL(at::Tensor, Boxed)::list_type::const_iterator&
+TORCH_ILIST_IMPL(at::Tensor, Boxed)::unwrap(
+    const c10::IListIterator<at::Tensor>& it
+) {
+  return it.payload_.boxed_iterator;
+}
+
+inline
+c10::detail::IListConstRef<at::Tensor>
+TORCH_ILIST_IMPL(at::Tensor, Boxed)::iterator_get(
+    const list_type::const_iterator& it
+) {
+  return (*it).get().toTensor();
+}
+
+inline
+const TORCH_ILIST_IMPL(at::OptionalTensorRef, Unboxed)::list_type&
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Unboxed)::unwrap(
+    const c10::IList<at::OptionalTensorRef>& ilist
+) {
+  return ilist.payload_.unboxed;
+}
+
+inline
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Unboxed)::list_type::const_iterator&
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Unboxed)::unwrap(
+    c10::IListIterator<at::OptionalTensorRef>& it
+) {
+  return it.payload_.unboxed_iterator;
+}
+
+inline
+const TORCH_ILIST_IMPL(at::OptionalTensorRef, Unboxed)::list_type::const_iterator&
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Unboxed)::unwrap(
+    const c10::IListIterator<at::OptionalTensorRef>& it
+) {
+  return it.payload_.unboxed_iterator;
+}
+
+inline
+c10::detail::IListConstRef<at::OptionalTensorRef>
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Unboxed)::iterator_get(
+    const list_type::const_iterator& it
+) {
+    return *it;
+}
+
+inline
+const TORCH_ILIST_IMPL(at::OptionalTensorRef, Boxed)::list_type&
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Boxed)::unwrap(
+    const c10::IList<at::OptionalTensorRef>& ilist
+) {
+  return *ilist.payload_.boxed;
+}
+
+inline
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Boxed)::list_type::const_iterator&
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Boxed)::unwrap(
+    c10::IListIterator<at::OptionalTensorRef>& it
+) {
+  return it.payload_.boxed_iterator;
+}
+
+inline
+const TORCH_ILIST_IMPL(at::OptionalTensorRef, Boxed)::list_type::const_iterator&
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Boxed)::unwrap(
+    const c10::IListIterator<at::OptionalTensorRef>& it
+) {
+  return it.payload_.boxed_iterator;
+}
+
+inline
+c10::detail::IListConstRef<at::OptionalTensorRef>
+TORCH_ILIST_IMPL(at::OptionalTensorRef, Boxed)::iterator_get(
+    const list_type::const_iterator& it
+) {
+  const auto& ivalue = (*it).get();
+  return (ivalue.isNone()) ? at::OptionalTensorRef{} : ivalue.toTensor();
+}
+
+namespace at {
+// [Note: ITensorList]
```

Comment:
Are you going to add a description here of why we have / need ITensorlist (and IOptTensorRefList)? Maybe summarizing some of the discussion from the gh issue so people can quickly come to this file and understand why we have these two new types.

Separately - you have a bunch of good comments defined in `IList.h`. Can you add a bit more on (a) how to add a new tensor type (Tensor, OptionalTensorRef, _____), and (b) how to add a new tag (IListTag::unboxed, IListTag::boxed, _____)? For example, it looks like we need to define a new `IListTagImpl` for the full cross product of [tags] x [tensor types]

Reply:
> Are you going to add a description here of why we have / need ITensorlist (and IOptTensorRefList)?

I wasn't planning to do that. But, I guess it would be better to do so.

> Can you add a bit more on (a) how to add a new tensor type (Tensor, OptionalTensorRef, _____), and (b) how to add a new tag (IListTag::unboxed, IListTag::boxed, _____)? For example, it looks like we need to define a new `IListTagImpl` for the full cross product of [tags] x [tensor types]

Sure, I will do that. Yeah, currently we need to do that, which is quite verbose. The reason for that is that `list_type` for the boxed and unboxed versions might change, depending on what is the parameter type of `IList` (see below).

```c++
/* IList<at::Tensor> */
IListTagImpl<at::Tensor, IListTag::Unboxed>::list_type == ArrayRef<at::Tensor>
IListTagImpl<at::Tensor, IListTag::Boxed>::list_type == List<at::Tensor>

/* IList<at::OptionalTensorRef> */
IListTagImpl<at::OptionalTensorRef, IListTag::Unboxed>::list_type == ArrayRef<at::OptionalTensorRef>
IListTagImpl<at::OptionalTensorRef, IListTag::Boxed>::list_type == List<optional<at::Tensor>>
```

That said, I'll try to play with it a little bit more and see if I can get a better design.